### PR TITLE
Track memory usage

### PR DESCRIPTION
**Pull request by arkpar**
This allows reusing module instances in substrate. Memory used by each call can be tracked and cleared between calls.

### CODE REVIEW

**Comment by arkpar** [comment]
The idea is to "snapshot" memory region that had been set up by the data segment initialization and clear all **used** memory outside that region after each call. Sice e.g. rust compiler allocates one megabyte for stack that grows backwards and we want to only clear that small portion of it that was actually used.

File: src/memory.rs [hunk 1]
```diff
@@ -148,6 +150,16 @@ impl MemoryInstance {
         self.maximum
     }
 
+    /// Returns lowest offset ever written or `u32::max_value()` if none.
+    pub fn lowest_used(&self) -> u32 {
+        self.lowest_used.get()
```
[the remark this answers is not in the document]
Comment (arkpar) [reply src/memory.rs]: Updated

File: src/memory.rs [hunk 2]
```diff
@@ -474,7 +497,12 @@ impl MemoryInstance {
     ///
     /// Returns `Err` if the specified region is out of bounds.
     pub fn zero(&self, offset: usize, len: usize) -> Result<(), Error> {
-        self.clear(offset, 0, len)
+        if (offset + len) >= self.buffer.borrow().len() {
```
[the remark this answers is not in the document]
Comment (arkpar) [reply src/memory.rs]: Reverted this change.

**Comment by pepyakin** [comment]
Ok, I see. This scheme could work altough it implies particular memory layout (e.g. `--stack-first`) and adds some amount of non-determinism around trapping on OOB.

I think I'm ok with merging this one, although API looks really constrained to this use case. In future I think we might want to experiment with CoW techniques after we get #45

**Pull request merged**